Repurposing `guides/debugging.md` page (#536)
withastro/docs: src/pages/en/guides/troubleshooting.md

**Solution**: Write your imports and exports at the top of your component script.

**Status**: Current limitation; fix is being worked on.

💡 Not sure that this is your problem? Check to see if anyone else has reported [this issue](https://github.com/withastro/astro/issues?q=is%3Aissue+is%3Aopen+Transform+failed+with+*+error)!

### Cannot use import statement outside a module

In Astro components, `<script>` tags are hoisted and loaded as [JS modules](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules) by default. If you have included the [`is:inline` directive](/en/reference/directives-reference/#isinline) or any other attribute in your tag, this default behavior is removed.

**Solution**: If you have added any attributes to your `<script>` tag, you must also add the `type="module"` attribute to be able to use import statements.

**Status**: Expected Astro behavior, as intended.

💡 Not sure that this is your problem? Check to see if anyone else has reported [this issue](https://github.com/withastro/astro/issues?q=is%3Aissue+is%3Aopen+Cannot+use+import+statement)!

> ### Unable to render component
> 
> This is indicates an error in a component you have imported and are using in your Astro template.
> 
> **Solution**: Check the appropriate documentation for your [Astro](/en/core-concepts/astro-components/) or [UI framework](/en/core-concepts/framework-components/) component. Consider opening an Astro starter template from [astro.new](https://astro.new) and troubleshooting just your component in a minimal Astro project.
> 
> > This can also be caused by attempting to access the `window` or `document` object at render time.
> 
> By default, Astro will render your component [isomorphically](https://en.wikipedia.org/wiki/Isomorphic_JavaScript), meaning it runs on the server where the browser runtime is not available. You can disable this pre-render step using [the `client:only` directive](/en/reference/directives-reference/#clientonly).
> 
> **Solution**: Try to access those objects after rendering (ex: [`useEffect()`](https://reactjs.org/docs/hooks-reference.html#useeffect) in React or [`onMounted()`](https://vuejs.org/api/composition-api-lifecycle.html#onmounted) in Vue and Svelte)
> 
> **Status**: Expected Astro behavior, as intended.

Reviewer: ```suggestion
### Unable to render component

This is indicates an error in a component you have imported and are using in your Astro template.

#### Common cause

This can be caused by attempting to access the `window` or `document` object at render time. By default, Astro will render your component [isomorphically](https://en.wikipedia.org/wiki/Isomorphic_JavaScript), meaning it runs on the server where the browser runtime is not available. You can disable this pre-render step using [the `client:only` directive](/en/reference/directives-reference/#clientonly).

**Solution**: Try to access those objects after rendering (ex: [`useEffect()`](https://reactjs.org/docs/hooks-reference.html#useeffect) in React or [`onMounted()`](https://vuejs.org/api/composition-api-lifecycle.html#onmounted) in Vue and Svelte)

**Status**: Expected Astro behavior, as intended.

#### Not that?

**Solution**: Check the appropriate documentation for your [Astro](/en/core-concepts/astro-components/) or [UI framework](/en/core-concepts/framework-components/) component. Consider opening an Astro starter template from [astro.new](https://astro.new) and troubleshooting just your component in a minimal Astro project.
```

### My component is not interactive

If your component is rendering (see above) but is not responding to user interaction, then you may be missing a [`client:*` directive](/en/reference/directives-reference/#client-directives) to hydrate your component.

By default, a [UI Framework component is not hydrated in the client](/en/core-concepts/framework-components/#hydrating-interactive-components). If no `client:*` directive is provided, its HTML is rendered onto the page without JavaScript.

Note: [Astro components](/en/core-concepts/astro-components/) are HTML-only templating components with no client-side runtime. But, you can use a `<script>` tag in your Astro component template to send JavaScript to the browser that executes in the global scope.

### Cannot find package 'X'

See [Astro's integration guide](/en/guides/integrations-guide/) for instructions on adding framework renderers, CSS tools and other packages to Astro.

You may need to install peer dependencies for some integrations. If you see a “missing peer dependencies” warning, you can follow the instructions for  [handling dependencies](/en/guides/integrations-guide/#handling-integration-dependencies).

> ### `Astro.glob()`'s "No matches found"

Reviewer: Nitpicking, so you _can_ ignore, but the possessive seems odd here

```suggestion
### `Astro.glob()` "No matches found" error
```

Reviewer: **_or_** 

```suggestion
### `Astro.glob()` error: "No matches found"
```

Reviewer: **_or_**

something else?

> ### `Astro.glob()`'s "No matches found"

Reviewer 2: ```suggestion
### `Astro.glob()` - no matches found
```

When using `Astro.glob()` to import files, be sure to use the correct glob syntax that will match all the files you need.

> 

Reviewer 2: ```suggestion

#### Filepaths

```

> For example, use `src/components/**/*.js` to import both of the following files:

Reviewer: The glob pattern _must_ be a relative path right? So actually `src/components/**/*.js` would not work. You need `../components/**/*.js` or something I think? Which makes the example trickier to show. Maybe:

```suggestion
For example, use `../components/**/*.js` in `src/pages/index.astro` to import both of the following files:
```

Reviewer 2: So my question is, is this supposed to be demonstrating the proper relative path that goes INSIDE THE Astro.glob()

`Astro.glob('../components/**/*.js')` 

And if so, does the line make sense as: 

For example, use `Astro.glob('../components/**/*.js')` to import both of the following files:

Reviewer: Because the path is relative, this pattern will only work in specific files.

The line does make sense, but if the file that contained it were nested differently alongside the `components` directory, it wouldn’t work.

- `Astro.glob('../components/**/*.js')` in `src/pages/index.astro` works
- `Astro.glob('../components/**/*.js')` in `src/pages/about/index.astro` doesn’t

Reviewer 2: Got it. OK, then I'll go with your initial suggestion for simplicity! It gets the point across, and isn't meant to be comprehensive!

- `src/components/MyComponent.js`
- `src/components/includes/MyOtherComponent.js`

> > Note that dynamic variables and string interpolation in `Astro.glob()` is not supported.

Reviewer 2: ```suggestion
#### Supported Values

 `Astro.glob()` does not support dynamic variables and string interpolation.
```

This is not a bug in Astro. It is due to a limitation of [Vite's `import.meta.glob()` function](https://vitejs.dev/guide/features.html#glob-import) which only supports static string literals.

A common workaround is to instead import a larger set of files that includes all the files you need using `Astro.glob()`, then filter them:

```astro
---
// src/components/featured.astro
const { postSlug } = Astro.props
const pathToMyFeaturedPost = `src/pages/blog/${postSlug}.md`

const posts = await Astro.glob('../pages/blog/*.md');

> const myFeaturedPost = posts.find(post => post.file.indexOf(pathToMyFeaturedPost) !== -1);

Reviewer 2: ```suggestion
const myFeaturedPost = posts.find(post => post.file.includes(pathToMyFeaturedPost));
```

> > Note that dynamic variables and string interpolation in `Astro.glob()` is not supported.
> 
> This is not a bug in Astro. It is due to a limitation of [Vite's `import.meta.glob()` function](https://vitejs.dev/guide/features.html#glob-import) which only supports static string literals.
> 
> A common workaround is to instead import a larger set of files that includes all the files you need using `Astro.glob()`, then filter them:
> 
> ```astro
> ---
> // src/components/featured.astro
> const { postSlug } = Astro.props
> const pathToMyFeaturedPost = `src/pages/blog/${postSlug}.md`
> 
> const posts = await Astro.glob('../pages/blog/*.md');
> const myFeaturedPost = posts.find(post => post.file.indexOf(pathToMyFeaturedPost) !== -1);
> ---
> 
> <p>
>     Take a look at my favorite post, <a href={myFeaturedPost.url}>{myFeaturedPost.frontmatter.title}</a>!
> </p>
> ```

Reviewer: Should the blockquote maybe come at the end of this section rather than in the middle?

I’ve also updated the `posts.find` line for a slightly more readable alternative.

````suggestion
This is not a bug in Astro. It is due to a limitation of [Vite's `import.meta.glob()` function](https://vitejs.dev/guide/features.html#glob-import) which only supports static string literals.

A common workaround is to instead import a larger set of files that includes all the files you need using `Astro.glob()`, then filter them:

```astro
---
// src/components/featured.astro
const { postSlug } = Astro.props
const pathToMyFeaturedPost = `src/pages/blog/${postSlug}.md`

const posts = await Astro.glob('../pages/blog/*.md');
const myFeaturedPost = posts.find(post => post.file.includes(pathToMyFeaturedPost));
---

<p>
    Take a look at my favorite post, <a href={myFeaturedPost.url}>{myFeaturedPost.frontmatter.title}</a>!
</p>
```

> Note that dynamic variables and string interpolation in `Astro.glob()` is not supported.
````

Reviewer 2: I think "This is not a bug"  and the workaround example refers to that call out, so I'm not sure that it makes sense to move that below?

Is the callout distracting, or does it seem out of place?

Reviewer: OH! If the blockquote and the following stuff is _connected_ I think they should either both be in the blockquote or the blockquote shouldn’t be there. I see the blockquote as separating out everything inside from what’s outside.

Reviewer 2: Yeah, look at the changes I commited to the branch and see if that reads better?

Reviewer: And… I just saw you already fixed this 💜
